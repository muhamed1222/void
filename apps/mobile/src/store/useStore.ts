import { create } from 'zustand';
import { persist } from 'zustand/middleware';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { TaskItem, Message } from '@/domain/types';

// Create a storage interface for zustand persist
const customStorage = {
  getItem: async (key: string) => {
    try {
      const value = await AsyncStorage.getItem(key);
      return value ? JSON.parse(value) : null;
    } catch (e) {
      console.error('Error getting item from storage:', e);
      return null;
    }
  },
  setItem: async (key: string, value: any) => {
    try {
      const valueString = JSON.stringify(value);
      await AsyncStorage.setItem(key, valueString);
    } catch (e) {
      console.error('Error setting item in storage:', e);
    }
  },
  removeItem: async (key: string) => {
    try {
      await AsyncStorage.removeItem(key);
    } catch (e) {
      console.error('Error removing item from storage:', e);
    }
  },
};

// Define the store state
interface AppState {
  // User ID state
  userId: string | null;
  hasAssignedId: boolean;

  // Today screen state
  contextualInstruction: string;
  todayTasks: TaskItem[];
  coachMessages: Message[];
  streak: number;
  disciplineScore: number;
  focusTime: number; // minutes today

  // Actions for Today screen
  addTask: (task: TaskItem) => void;
  updateTask: (id: string, patch: Partial<TaskItem>) => void;
  removeTask: (id: string) => void;
  completeTask: (id: string) => void;
  setContextualInstruction: (instruction: string) => void;
  addCoachMessage: (message: Message) => void;
  setStreak: (value: number) => void;
  setDisciplineScore: (value: number) => void;
  setFocusTime: (minutes: number) => void;

  // User ID actions
  setUserId: (id: string) => void;
  setHasAssignedId: (assigned: boolean) => void;

  // Console state
  consoleMessages: string[];
  addConsoleMessage: (message: string) => void;

  // UI flags
  isTimerRunning: boolean;
  setTimerRunning: (running: boolean) => void;
}

// Create the store with persistence
export const useStore = create<AppState>()(
  persist(
    (set, get) => ({
      // Initial state
      userId: null,
      hasAssignedId: false,
      contextualInstruction: '',
      todayTasks: [],
      streak: 0,
      disciplineScore: 0,
      focusTime: 0,
      coachMessages: [],
      consoleMessages: [],
      isTimerRunning: false,
      
      // Actions
      addTask: (task) => set((state) => ({
        todayTasks: [...state.todayTasks, task],
      })),
      updateTask: (id, patch) => set((state) => ({
        todayTasks: state.todayTasks.map((t) => (t.id === id ? { ...t, ...patch } : t)),
      })),
      removeTask: (id) => set((state) => ({
        todayTasks: state.todayTasks.filter((t) => t.id !== id),
      })),
      completeTask: (id) => set((state) => ({
        todayTasks: state.todayTasks.map((t) => (t.id === id ? { ...t, completed: true, completedAt: new Date().toISOString() } : t)),
      })),
      setContextualInstruction: (instruction) => set({ contextualInstruction: instruction }),
      addCoachMessage: (message) => set((state) => {
        // Check if a message with the same ID already exists
        const messageExists = state.coachMessages.some(msg => msg.id === message.id);
        
        // Only add the message if it doesn't already exist
        if (!messageExists) {
          return { coachMessages: [...state.coachMessages, message] };
        }
        
        // If message already exists, return the current state unchanged
        return { coachMessages: state.coachMessages };
      }),
      setStreak: (value) => set({ streak: value }),
      setDisciplineScore: (value) => set({ disciplineScore: value }),
      setFocusTime: (minutes) => set({ focusTime: minutes }),
      
      addConsoleMessage: (message) => set((state) => ({
        consoleMessages: [...state.consoleMessages, message]
      })),
      
      setTimerRunning: (running) => set({ isTimerRunning: running }),

      // User ID actions
      setUserId: (id) => set({ userId: id }),
      setHasAssignedId: (assigned) => set({ hasAssignedId: assigned }),
    }),
    {
      name: 'white-room-storage',
      storage: customStorage,
      version: 1, // STATE_VERSION for migrations
    }
  )
);