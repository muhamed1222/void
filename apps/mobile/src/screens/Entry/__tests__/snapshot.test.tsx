import React from 'react';
import { render } from '@testing-library/react-native';
import { EntryScreen } from '../index';
import { useStore } from '@/store/useStore';
import { NavigationContainer } from '@react-navigation/native';

// Mock the store
jest.mock('@/store/useStore', () => ({
  useStore: jest.fn(),
}));

// Mock navigation
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({
    reset: jest.fn(),
  }),
}));

// Mock theme hook
jest.mock('@/theming/hooks', () => ({
  useTheme: () => ({
    colors: {
      background: '#FFFFFF',
      text: '#000000',
      textSecondary: '#666666',
    },
  }),
}));

describe('EntryScreen Snapshot', () => {
  it('renders correctly with existing user ID', () => {
    (useStore as jest.Mock).mockReturnValue({
      userId: 'User#123',
      hasAssignedId: true,
      setUserId: jest.fn(),
      setHasAssignedId: jest.fn(),
    });

    const tree = render(
      <NavigationContainer>
        <EntryScreen />
      </NavigationContainer>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});