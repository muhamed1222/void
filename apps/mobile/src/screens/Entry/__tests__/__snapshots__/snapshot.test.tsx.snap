// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryScreen Snapshot renders correctly with existing user ID 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "padding": 16,
      },
      {
        "backgroundColor": "#FFFFFF",
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "fontSize": 16,
          "marginBottom": 8,
        },
        {
          "color": "#000000",
          "fontFamily": "monospace",
        },
      ]
    }
  >
    Assigning ID...
  </Text>
  <Text
    style={
      [
        {
          "fontSize": 16,
          "marginBottom": 8,
        },
        {
          "color": "#000000",
          "fontFamily": "monospace",
        },
      ]
    }
  >
    User#123
  </Text>
  <Text
    style={
      [
        {
          "fontSize": 16,
          "marginBottom": 24,
        },
        {
          "color": "#000000",
          "fontFamily": "monospace",
        },
      ]
    }
  >
    Допуск предоставлен.
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
        "padding": 12,
      }
    }
  >
    <Text
      style={
        [
          {
            "fontSize": 16,
          },
          {
            "color": "#000000",
            "fontFamily": "monospace",
          },
        ]
      }
    >
      [ Войти ]
    </Text>
  </View>
</View>
`;
